Sealer/sealer.cc

```diff
@@ -603,38 +602,48 @@ void kRun(const ConnectPairsParams& params,
 
 	printLog(logStream, "Flanks inserted into k run = " + IntToString(flanks.size()) + "\n");
 
-	for (read1_it = flanks.begin(); read1_it != flanks.end();) {
-		success = false;
+	int counter = 0;
+	vector<map<FastaRecord, map<FastaRecord, Gap> >::iterator> flanks_closed;
+#pragma omp parallel private(read1_it, read2_it) firstprivate(counter)
+	for (read1_it = flanks.begin(); read1_it != flanks.end(); read1_it++) {
 		FastaRecord read1 = read1_it->first;
-		for (read2_it = flanks[read1].begin(); read2_it != flanks[read1].end(); read2_it++) {
+		bool success = false;
+		for (read2_it = flanks[read1].begin(); read2_it != flanks[read1].end(); read2_it++, counter++) {
+			#if _OPENMP
+			if (counter % omp_get_num_threads() == omp_get_thread_num())
+				continue;
+			#endif
 			FastaRecord read2 = read2_it->first;
 
 			int startposition = read2_it->second.gapStart();
 			string tempSeq = merge(g, k, read2_it->second, read1, read2, params, g_count, traceStream);
 			if (!tempSeq.empty()) {
 				success = true;
+#pragma omp critical (allmerged)
 				allmerged[read1.id.substr(0,read1.id.length()-2)][startposition]
 					= ClosedGap(read2_it->second, tempSeq);
-//#pragma omp atomic
-				gapsclosed++;
-//#pragma omp atomic
+#pragma omp atomic
 				uniqueGapsClosed++;
-				if (gapsclosed % 100 == 0)
+#pragma omp critical (gapsclosed)
+				if (++gapsclosed % 100 == 0)
 					printLog(logStream, IntToString(gapsclosed) + " gaps closed so far\n");
 
-				if (!opt::gapfilePath.empty()) {
+				if (!opt::gapfilePath.empty())
+#pragma omp critical (gapfilePath)
 					gapStream << ">" << read1.id.substr(0,read1.id.length()-2)
 						  << "_" << read2_it->second.gapStart() << "-" << read2_it->second.gapEnd()
-						  << " LN:i:" << tempSeq.length() << '\n';
-					gapStream << tempSeq << '\n';
-				}
+						  << " LN:i:" << tempSeq.length() << '\n' << tempSeq << '\n';
 			}
 		}
 		if (success) {
-			flanks.erase(read1_it++);
+#pragma omp critical (flanks_closed)
+			flanks_closed.push_back(read1_it);
 		}
-		else
-			read1_it++;
+	}
+
+	for (vector<map<FastaRecord, map<FastaRecord, Gap> >::iterator>::iterator it = flanks_closed.begin();
+			it != flanks_closed.end(); it++) {
+		flanks.erase(*it);
```

Comment:
I'm concerned that this code may not be valid semantically. It's a sticky part of C++, but I don't believe that you're permitted to `erase` an iterator that your actively using for iteration. It's different for each data structure though. you may need something like
```c++
for (vector<map<FastaRecord, map<FastaRecord, Gap> >::iterator>::iterator
 			it = flanks_closed.begin();
 			it != flanks_closed.end();)
 	it = flanks.erase(it);
```
You'll find posts on the topic when you Google for it.

Reply:
What is the purpose though of erasing these elements from `flank`?

Reply:
Still needs addressing.

Reply:
There are 2 data structures here - flanks_closed and flanks. I am iterating flanks_closed that holds iterators that belong to flanks, and delete the data they point to. This should be valid, as I am not traversing flanks here, no? Before this pull request, this was done while traversing flanks and closing gaps (but in a safe way, by moving the iterator, then deleting what it previously pointed to), but I had to change it because there are now multiple threads working with flanks, so it's only safe to alter flanks after the work is done.

The reason, functionality wise, for erasing flanks, is because after kRun finishes, it proceeds to the next value of k and if they weren't erased, this new value of k could potentially close the same gaps from the previous run. At least that is my understanding, but in any case, the original code of Sealer was erasing flanks that belong to closed gaps, so I wanted to keep the functionality.

Reply:
So that means the original Sealer, before this pull request, was potentially not dealing with it safely either, as it deleted iterators on the go. Sounds like flagging processed flanks would be the safe way to do it, then.

Reply:
Ah, I scanned the code too quickly! I missed that `flanks_closed` is a new data structure, and that it's a vector of iterators. I'm okay with the code as is. For a future PR to tidy up the code, I'd prefer that flanks were a vector of tuples, rather than a map of maps as it is now.